File: Herencia/Ejercicio5/Empleado.java
package Herencia.Ejercicio5;

import java.util.ArrayList;

public class Empleado {
    private String nombre, apellido;
    private int antiguedad, salarioBase;

    public Empleado(String nombre, String apellido, int antiguedad, int salarioBase) {
        this.nombre = nombre;
        this.apellido = apellido;
        this.antiguedad = antiguedad;
        this.salarioBase = salarioBase;
    }

    public String getNombre() {
        return nombre;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public String getApellido() {
        return apellido;
    }

    public void setApellido(String apellido) {
        this.apellido = apellido;
    }

    public int getAntiguedad() {
        return antiguedad;
    }

    public void setAntiguedad(int antiguedad) {
        this.antiguedad = antiguedad;
    }

    public int getSalarioBase() {
        return salarioBase;
    }

    public void setSalarioBase(int salarioBase) {
        this.salarioBase = salarioBase;
    }

    public double calcularSalario() {
        return salarioBase + (salarioBase * 0.05 * antiguedad);
    }

    public void mostrarInfo() {
        System.out.printf("Nombre: %s %s | Antigüedad: %d años | Salario: $%.2f%n",
                nombre, apellido, antiguedad, calcularSalario());
    }
}

// Clase Gerente
class Gerente extends Empleado {
    private String departamento;
    private int bonoGerencial;

    public Gerente(String nombre, String apellido, int antiguedad, int salarioBase, String departamento, int bonoGerencial) {
        super(nombre, apellido, antiguedad, salarioBase);
        this.departamento = departamento;
        this.bonoGerencial = bonoGerencial;
    }

    public String getDepartamento() {
        return departamento;
    }

    public void setDepartamento(String departamento) {
        this.departamento = departamento;
    }

    public int getBonoGerencial() {
        return bonoGerencial;
    }

    public void setBonoGerencial(int bonoGerencial) {
        this.bonoGerencial = bonoGerencial;
    }

    @Override
    public double calcularSalario() {
        return super.calcularSalario() + bonoGerencial;
    }

    @Override
    public void mostrarInfo() {
        System.out.printf("Gerente: %s %s | Departamento: %s | Salario: $%.2f%n",
                getNombre(), getApellido(), departamento, calcularSalario());
    }
}

// Clase Desarrollador
class Desarrollador extends Empleado {
    private String lenguajeProgramacion;
    private int horasExtras;

    public Desarrollador(String nombre, String apellido, int antiguedad, int salarioBase, String lenguajeProgramacion, int horasExtras) {
        super(nombre, apellido, antiguedad, salarioBase);
        this.lenguajeProgramacion = lenguajeProgramacion;
        this.horasExtras = horasExtras;
    }

    public String getLenguajeProgramacion() {
        return lenguajeProgramacion;
    }

    public void setLenguajeProgramacion(String lenguajeProgramacion) {
        this.lenguajeProgramacion = lenguajeProgramacion;
    }

    public int getHorasExtras() {
        return horasExtras;
    }

    public void setHorasExtras(int horasExtras) {
        this.horasExtras = horasExtras;
    }

    @Override
    public double calcularSalario() {
        return super.calcularSalario() + horasExtras * 100;
    }

    @Override
    public void mostrarInfo() {
        System.out.printf("Desarrollador: %s %s | Lenguaje: %s | Salario: $%.2f%n",
                getNombre(), getApellido(), lenguajeProgramacion, calcularSalario());
    }
}

// Clase Main para probar las funcionalidades
class Main {
    public static void main(String[] args) {
        // Crear lista de empleados
        ArrayList<Gerente> gerentes = new ArrayList<>();
        ArrayList<Desarrollador> desarrolladores = new ArrayList<>();

        // Crear instancias de Gerente y Desarrollador
        Gerente g1 = new Gerente("Carlos", "Pérez", 10, 5000, "TI", 1500);
        Gerente g2 = new Gerente("María", "González", 5, 4500, "Ventas", 800);
        Gerente g3 = new Gerente("Juan", "López", 7, 4700, "Marketing", 1200);

        Desarrollador d1 = new Desarrollador("Ana", "Fernández", 3, 3000, "Java", 12);
        Desarrollador d2 = new Desarrollador("Luis", "Gómez", 2, 2800, "Python", 8);
        Desarrollador d3 = new Desarrollador("Pedro", "Martínez", 4, 3200, "JavaScript", 15);

        // Agregar empleados a las listas
        gerentes.add(g1);
        gerentes.add(g2);
        gerentes.add(g3);

        desarrolladores.add(d1);
        desarrolladores.add(d2);
        desarrolladores.add(d3);

        // b) Mostrar salario de cada gerente y desarrollador
        System.out.println("\nSalarios de Gerentes:");
        for (Gerente g : gerentes) {
            g.mostrarInfo();
        }

        System.out.println("\nSalarios de Desarrolladores:");
        for (Desarrollador d : desarrolladores) {
            d.mostrarInfo();
        }

        // c) Mostrar gerentes con bono gerencial mayor a 1000
        System.out.println("\nGerentes con bono mayor a 1000:");
        for (Gerente g : gerentes) {
            if (g.getBonoGerencial() > 1000) {
                g.mostrarInfo();
            }
        }

        // d) Mostrar desarrolladores con más de 10 horas extras
        System.out.println("\nDesarrolladores con más de 10 horas extras:");
        for (Desarrollador d : desarrolladores) {
            if (d.getHorasExtras() > 10) {
                d.mostrarInfo();
            }
        }
    }
}

File: Herencia/Ejercicio5/Empleado.py
class Persona:
    def __init__(self, nombre, apellido, fecha_nac, ci, celular):
        self.nombre = nombre
        self.apellido = apellido
        self.fecha_nac = fecha_nac
        self.ci = ci
        self.celular = celular
    
    def get_apellido(self):
        return self.apellido
    
    def get_nombre(self):
        return self.nombre
    
    def get_edad(self):
        año_nacimiento = int(self.fecha_nac.split("/")[2])  # Extrae el año como entero
        return 2025 - año_nacimiento  # Suponiendo el año actual como 2025
    
    def mostrar(self):
        print(f"Nombre: {self.nombre} | Apellido: {self.apellido} | Fecha de Nacimiento: {self.fecha_nac} | CI: {self.ci} | Celular: {self.celular}")


class Estudiante(Persona):
    def __init__(self, nombre, apellido, fecha_nac, ci, celular, ru, semestre, fecha_ingreso):
        super().__init__(nombre, apellido, fecha_nac, ci, celular)
        self.ru = ru
        self.semestre = semestre
        self.fecha_ingreso = fecha_ingreso
    
    def mostrar(self):
        super().mostrar()
        print(f" | RU: {self.ru} | Semestre: {self.semestre} | Fecha de Ingreso: {self.fecha_ingreso}")


class Docente(Persona):
    def __init__(self, nombre, apellido, fecha_nac, ci, celular, profesion, especialidad, nit, sexo):
        super().__init__(nombre, apellido, fecha_nac, ci, celular)
        self.profesion = profesion
        self.especialidad = especialidad
        self.nit = nit
        self.sexo = sexo
    
    def get_profesion(self):
        return self.profesion
    
    def get_sexo(self):
        return self.sexo
    
    def mostrar(self):
        super().mostrar()
        print(f" | Profesión: {self.profesion} | Especialidad: {self.especialidad} | NIT: {self.nit}")


# Clase Main para pruebas
if __name__ == "__main__":
    # Lista de estudiantes y docentes
    estudiantes = []
    docentes = []

    # Crear estudiantes
    estudiantes.append(Estudiante("Carlos", "Pérez", "15/06/1998", 1234567, 789654, 1001, 8, "01/03/2020"))
    estudiantes.append(Estudiante("Ana", "López", "20/08/2001", 7654321, 321987, 1002, 5, "01/03/2022"))
    estudiantes.append(Estudiante("Juan", "González", "30/09/1995", 4567891, 852741, 1003, 10, "01/03/2018"))

    # Crear docentes
    docentes.append(Docente("Mario", "Fernández", "05/07/1980", 1122334, 123456, "Ingeniero", "Software", 556677, "Masculino"))
    docentes.append(Docente("Luis", "Gómez", "12/11/1975", 2233445, 987654, "Ingeniero", "Redes", 112233, "Masculino"))
    docentes.append(Docente("Laura", "Pérez", "25/04/1985", 3344556, 456123, "Licenciada", "Matemáticas", 778899, "Femenino"))

    # a) Mostrar estudiantes mayores de 25 años
    print("\n--- Estudiantes mayores de 25 años ---")
    for est in estudiantes:
        if est.get_edad() > 25:
            est.mostrar()

    # b) Mostrar al docente Ingeniero masculino más mayor
    mayor_ingeniero = None
    for doc in docentes:
        if doc.get_profesion() == "Ingeniero" and doc.get_sexo() == "Masculino":
            if mayor_ingeniero is None or doc.get_edad() > mayor_ingeniero.get_edad():
                mayor_ingeniero = doc
    
    if mayor_ingeniero is not None:
        print("\n--- Docente Ingeniero Masculino Más Mayor ---")
        mayor_ingeniero.mostrar()

    # c) Mostrar estudiantes y docentes con el mismo apellido
    print("\n--- Personas con el mismo apellido ---")
    for est in estudiantes:
        for doc in docentes:
            if est.get_apellido() == doc.get_apellido():
                print(f"Coincidencia de apellido: {est.get_apellido()}")
                est.mostrar()
                doc.mostrar()